File: examples/python/05-extract-structured/main.py
# Extract words with bounding boxes and tables from a PDF page.
# Run: python main.py document.pdf

import sys

from pdf_oxide import PdfDocument


def main():
    if len(sys.argv) < 2:
        print("Usage: python main.py <file.pdf>")
        sys.exit(1)

    path = sys.argv[1]
    doc = PdfDocument(path)
    print(f"Opened: {path}")

    page = 0

    # Extract words with position data
    words = doc.extract_words(page)
    print(f"\n--- Words (page {page + 1}) ---")
    for w in words[:20]:
        quoted = '"' + w["text"] + '"'
        print(f'{quoted:<20} '
              f'x={w["x0"]:<7.1f} y={w["y0"]:<7.1f} '
              f'x1={w["x1"]:<7.1f} y1={w["y1"]:<7.1f} '
              f'font={w["fontname"]}  size={w["size"]:.1f}')
    if len(words) > 20:
        print(f"... ({len(words) - 20} more words)")

    # Extract tables
    tables = doc.extract_tables(page)
    print(f"\n--- Tables (page {page + 1}) ---")
    if not tables:
        print("(no tables found)")
    for i, table in enumerate(tables):
        rows, cols = len(table), len(table[0]) if table else 0
        print(f"Table {i + 1}: {rows} rows x {cols} cols")
        for r, row in enumerate(table[:5]):
            cells = "  ".join(f'[{r},{c}] "{v}"' for c, v in enumerate(row[:6]))
            print(f"  {cells}")

if __name__ == "__main__":
    main()

File: examples/rust/05-extract-structured/main.rs
// Extract words with bounding boxes and tables from a PDF page.
// Run: cargo run --example extract_structured -- document.pdf

use pdf_oxide::PdfDocument;
use std::env;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let path = env::args().nth(1).expect("Usage: extract_structured <file.pdf>");
    let doc = PdfDocument::open(&path)?;
    println!("Opened: {}", path);

    let page = 0;

    // Extract words with position data
    let words = doc.extract_words(page)?;
    println!("\n--- Words (page {}) ---", page + 1);
    for w in words.iter().take(20) {
        println!(
            "{:20} x={:<7.1} y={:<7.1} w={:<7.1} h={:<7.1} font={} size={:.1}",
            format!("\"{}\"", w.text),
            w.x, w.y, w.width, w.height, w.font_name, w.font_size
        );
    }
    if words.len() > 20 {
        println!("... ({} more words)", words.len() - 20);
    }

    // Extract tables
    let tables = doc.extract_tables(page)?;
    println!("\n--- Tables (page {}) ---", page + 1);
    if tables.is_empty() {
        println!("(no tables found)");
    }
    for (i, table) in tables.iter().enumerate() {
        println!("Table {}: {} rows x {} cols", i + 1, table.rows, table.cols);
        for r in 0..table.rows.min(5) {
            let row_str: Vec<String> = (0..table.cols.min(6))
                .map(|c| format!("[{},{}] \"{}\"", r, c, table.cells[r][c]))
                .collect();
            println!("  {}", row_str.join("  "));
        }
    }

    Ok(())
}
